File: frontend/src/api/pricing.ts
import { del, get, patch, post } from './client'
import type { StorageType } from './billing'

// ── Types ──────────────────────────────────────────────────────────────────────

// PricingItem is one purchasable storage line item on a server tier, as
// managed on the admin pricing page.
export interface PricingItem {
  id: string
  server_id: string
  storage_type: StorageType
  bytes: number
  price_cents: number
  sort_order: number
  created_at: string
  updated_at: string
}

export type DiscountScope = 'server' | 'tier' | 'item'
export type DiscountMode = 'percent' | 'price'
export type NotifyGroup = 'all' | 'server' | 'server_tier'

// PricingDiscount is an admin-created markdown of one pricing target. Only
// one exists per exact target; the most specific scope wins per item.
export interface PricingDiscount {
  id: string
  scope: DiscountScope
  server_id: string
  storage_type?: StorageType
  item_id?: string
  mode: DiscountMode
  percent_off?: number
  price_cents?: number
  premium_only: boolean
  expires_at?: string
  notify_group?: NotifyGroup
  created_by: string
  created_at: string
}

// PricingServer is a row of the admin page's server picker.
export interface PricingServer {
  id: string
  name: string
  is_active: boolean
  tiers: StorageType[]
}

// ── Client-side effective price helpers (mirror models/pricing.go) ─────────────

const scopeRank: Record<DiscountScope, number> = { item: 0, tier: 1, server: 2 }

export function discountAppliesTo(d: PricingDiscount, item: PricingItem): boolean {
  if (d.scope === 'item') return d.item_id === item.id
  if (d.scope === 'tier') return d.server_id === item.server_id && d.storage_type === item.storage_type
  return d.server_id === item.server_id
}

export function discountActive(d: PricingDiscount, now = new Date()): boolean {
  return !d.expires_at || new Date(d.expires_at) > now
}

// resolveDiscount picks the most specific active discount for an item —
// item > tier > server, never stacking.
export function resolveDiscount(
  item: PricingItem,
  discounts: PricingDiscount[],
  now = new Date(),
): PricingDiscount | undefined {
  let best: PricingDiscount | undefined
  for (const d of discounts) {
    if (!discountActive(d, now) || !discountAppliesTo(d, item)) continue
    if (!best || scopeRank[d.scope] < scopeRank[best.scope]) best = d
  }
  return best
}

// discountedCents applies a discount to a base price: percent rounds half-up,
// reduced price clamps to the base so it can never mark an item up.
export function discountedCents(d: PricingDiscount, baseCents: number): number {
  if (d.mode === 'percent') {
    return Math.floor((baseCents * (100 - (d.percent_off ?? 0)) + 50) / 100)
  }
  return Math.min(d.price_cents ?? baseCents, baseCents)
}

// discountPercentFor is the badge percentage: the stored percent, or the
// per-item deduction derived from a reduced price.
export function discountPercentFor(d: PricingDiscount, baseCents: number): number {
  if (d.mode === 'percent') return d.percent_off ?? 0
  if (baseCents <= 0) return 0
  const off = baseCents - discountedCents(d, baseCents)
  return Math.round((off * 100) / baseCents)
}

// ── Admin endpoints ─────────────────────────────────────────────────────────────

export async function listPricingServers(): Promise<PricingServer[]> {
  const res = await get<{ servers: PricingServer[] }>('/admin/pricing/servers')
  return res.servers ?? []
}

export interface ServerPricing {
  items: PricingItem[]
  discounts: PricingDiscount[]
}

export function getServerPricing(serverId: string) {
  return get<ServerPricing>(`/admin/pricing?server_id=${encodeURIComponent(serverId)}`)
}

export function createPricingItem(input: {
  server_id: string
  storage_type: StorageType
  bytes: number
  price_cents: number
  sort_order?: number
}) {
  return post<{ item: PricingItem }>('/admin/pricing/items', input)
}

export function updatePricingItem(
  itemId: string,
  input: { bytes: number; price_cents: number; sort_order: number },
) {
  return patch<{ item: PricingItem }>(`/admin/pricing/items/${itemId}`, input)
}

export function deletePricingItem(itemId: string) {
  return del<void>(`/admin/pricing/items/${itemId}`)
}

export interface CreateDiscountInput {
  scope: DiscountScope
  server_id?: string
  storage_type?: StorageType
  item_id?: string
  mode: DiscountMode
  percent_off?: number
  price_cents?: number
  premium_only: boolean
  // ISO timestamp; omit for a discount that never expires.
  expires_at?: string
  // Email notification group; omit for no emails.
  notify?: NotifyGroup
}

export function createPricingDiscount(input: CreateDiscountInput) {
  return post<{ discount: PricingDiscount; recipients_notified: number }>(
    '/admin/pricing/discounts',
    input,
  )
}

export function deletePricingDiscount(discountId: string) {
  return del<void>(`/admin/pricing/discounts/${discountId}`)
}

// ── User-facing effective plans (storage upgrade modal) ─────────────────────────

// StoragePlanDiscount is the discount the backend already applied to a plan
// row, for the marked-down price display.
export interface StoragePlanDiscount {
  id: string
  scope: DiscountScope
  mode: DiscountMode
  percent: number
  expires_at?: string
  premium_only: boolean
}

// ServerStoragePlan is one purchasable line item priced for the calling user.
// An empty list from the endpoint means the server has no admin-managed
// pricing and the client should fall back to the legacy hardcoded plans.
export interface ServerStoragePlan {
  id: string
  storage_type: StorageType
  bytes: number
  price_cents: number
  effective_cents: number
  discount?: StoragePlanDiscount
}

export async function listServerStoragePlans(serverId: string): Promise<ServerStoragePlan[]> {
  const res = await get<{ items: ServerStoragePlan[] }>(
    `/billing/storage/plans?server_id=${encodeURIComponent(serverId)}`,
  )
  return res.items ?? []
}

File: api/models/pricing.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// PricingItem is one purchasable storage line item on a server tier —
// admin-edited via the pricing page and purchased by its UUID (the plan_id
// recorded on storage_orders). Servers without any items fall back to the
// legacy hardcoded plan table in routes/billing/plans.go.
type PricingItem struct {
	ID          uuid.UUID `json:"id"`
	ServerID    uuid.UUID `json:"server_id"`
	StorageType string    `json:"storage_type"` // "nvme" | "hdd"
	Bytes       int64     `json:"bytes"`
	PriceCents  int       `json:"price_cents"`
	SortOrder   int       `json:"sort_order"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// Pricing discount scopes and modes. Exactly one discount row exists per
// exact target (server / server+tier / item); the most specific applicable
// one wins at resolution time — discounts never stack.
const (
	DiscountScopeServer = "server"
	DiscountScopeTier   = "tier"
	DiscountScopeItem   = "item"

	DiscountModePercent = "percent"
	DiscountModePrice   = "price"
)

// PricingDiscount is an admin-created markdown of one pricing target.
// ServerID and StorageType are always populated (denormalized from the item
// for scope='item') so listings and recipient queries never need a join.
type PricingDiscount struct {
	ID          uuid.UUID  `json:"id"`
	Scope       string     `json:"scope"`
	ServerID    uuid.UUID  `json:"server_id"`
	StorageType *string    `json:"storage_type,omitempty"`
	ItemID      *uuid.UUID `json:"item_id,omitempty"`
	Mode        string     `json:"mode"`
	PercentOff  *int       `json:"percent_off,omitempty"`
	PriceCents  *int       `json:"price_cents,omitempty"`
	PremiumOnly bool       `json:"premium_only"`
	ExpiresAt   *time.Time `json:"expires_at,omitempty"`
	NotifyGroup *string    `json:"notify_group,omitempty"`
	CreatedBy   string     `json:"created_by"`
	CreatedAt   time.Time  `json:"created_at"`
}

// Active reports whether the discount is in effect at t (not expired).
func (d *PricingDiscount) Active(t time.Time) bool {
	return d.ExpiresAt == nil || d.ExpiresAt.After(t)
}

// AppliesTo reports whether the discount targets the given item, ignoring
// expiry and premium gating (callers filter those separately).
func (d *PricingDiscount) AppliesTo(item *PricingItem) bool {
	switch d.Scope {
	case DiscountScopeItem:
		return d.ItemID != nil && *d.ItemID == item.ID
	case DiscountScopeTier:
		return d.ServerID == item.ServerID &&
			d.StorageType != nil && *d.StorageType == item.StorageType
	case DiscountScopeServer:
		return d.ServerID == item.ServerID
	}
	return false
}

// discountScopeRank orders scopes most-specific-first for resolution.
var discountScopeRank = map[string]int{
	DiscountScopeItem:   0,
	DiscountScopeTier:   1,
	DiscountScopeServer: 2,
}

// ResolveDiscount picks the discount that applies to item from candidates:
// the most specific (item > tier > server) discount that is unexpired at now
// and — unless privileged (premium or admin user) — not premium-gated.
// Returns nil when none apply.
func ResolveDiscount(item *PricingItem, candidates []PricingDiscount, privileged bool, now time.Time) *PricingDiscount {
	var best *PricingDiscount
	for i := range candidates {
		d := &candidates[i]
		if !d.Active(now) || (d.PremiumOnly && !privileged) || !d.AppliesTo(item) {
			continue
		}
		if best == nil || discountScopeRank[d.Scope] < discountScopeRank[best.Scope] {
			best = d
		}
	}
	return best
}

// DiscountedPriceCents applies d to a base price. Percent mode rounds
// half-up; price mode clamps to the base price so a tier/server-wide reduced
// price can never mark an item up.
func (d *PricingDiscount) DiscountedPriceCents(baseCents int) int {
	switch d.Mode {
	case DiscountModePercent:
		if d.PercentOff == nil {
			return baseCents
		}
		return (baseCents*(100-*d.PercentOff) + 50) / 100
	case DiscountModePrice:
		if d.PriceCents == nil || *d.PriceCents > baseCents {
			return baseCents
		}
		return *d.PriceCents
	}
	return baseCents
}

// DiscountPercentFor is the percentage shown on the markdown badge: the
// stored percentage in percent mode, or the per-item deduction derived from
// the reduced price in price mode (rounded to the nearest whole percent).
func (d *PricingDiscount) DiscountPercentFor(baseCents int) int {
	if d.Mode == DiscountModePercent && d.PercentOff != nil {
		return *d.PercentOff
	}
	if baseCents <= 0 {
		return 0
	}
	off := baseCents - d.DiscountedPriceCents(baseCents)
	return (off*100 + baseCents/2) / baseCents
}
